File: src/ts/magma/api/collect/Iter.ts
import { Collector } from "../../../magma/api/collect/Collector";
import { Option } from "../../../magma/api/option/Option";
import { Result } from "../../../magma/api/result/Result";
export interface Iter<T> {
	collect<C>(collector: Collector<T, C>): C;
	map<R>(mapper: (arg0 : T) => R): Iter<R>;
	foldWithInitial<R>(initial: R, folder: (arg0 : R, arg1 : T) => R): R;
	foldWithMapper<R>(mapper: (arg0 : T) => R, folder: (arg0 : R, arg1 : T) => R): Option<R>;
	flatMap<R>(mapper: (arg0 : T) => Iter<R>): Iter<R>;
	next(): Option<T>;
	allMatch(predicate: (arg0 : T) => boolean): boolean;
	filter(predicate: (arg0 : T) => boolean): Iter<T>;
	anyMatch(predicate: (arg0 : T) => boolean): boolean;
	foldWithInitialToResult<R, X>(initial: R, folder: (arg0 : R, arg1 : T) => Result<R, X>): Result<R, X>;
}

File: src/java/magma/api/collect/iter/Iter.java
package magma.api.collect.iter;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;

public interface Iter<T> {
    <R> R fold(R initial, BiFunction<R, T, R> folder);

    Iter<T> filter(Predicate<T> predicate);

    <R> Iter<R> map(Function<T, R> mapper);

    <C> C collect(Collector<T, C> collector);
}
